lib/cext/ruby/ruby.h

```diff
@@ -391,11 +391,14 @@ ID SYM2ID(VALUE value);
 #endif
 #endif
 
+/* Truffle: Define our own RB_FLONUM_P
 #if USE_FLONUM
 #define RB_FLONUM_P(x) ((((int)(SIGNED_VALUE)(x))&RUBY_FLONUM_MASK) == RUBY_FLONUM_FLAG)
 #else
 #define RB_FLONUM_P(x) 0
 #endif
+*/
```

Comment:
Mmh, so we have USE_FLONUM as 1 currently.
I guess since we use unboxed doubles, it's fair enough to consider them all FLONUM (tagged doubles in MRI) and not full Float objects.
However, for `ruby_special_consts` just below it would make sense to have the USE_FLONUM=0 definitions (i.e. nil has object_id 4 for us).

Reply:
Okay. I wasn't really sure what was going on with the float object checks. There's also `RB_FLOAT_TYPE_P` we could use. But that was failing for me with the current definition. And I was having trouble understanding that macro since I can't readily tell what a special constant is supposed to be.

Reply:
`rb_special_const_p` is defined in ruby.c. That path should work too.
But it would be clearer and more efficient to just change `RB_FLOAT_TYPE_P` to be basically `obj.is_a? Float` for us.

Reply:
One idea: It might be possible to minimize the header change here a little more:
e.g.
```
# in truffle/ruby.h
int rb_tr_flonum_p(VALUE value);
```

```
#define RB_FLONUM_P(x) rb_tr_flonum_p(x)
```

Reply:
Sorry for not being clear. I could find where `rb_special_const_p` is defined for us, but a lot of the C extension API is only partially implemented or sometimes implemented incorrectly. So, I can't tell if our implementation is `rb_special_const_p` is correct because I also couldn't really tell from the MRI source what a special const is supposed to be.

Reply:
It's a tagged pointer in MRI basically.
For us it shouldn't really matter whether Float are considered tagged or not.